### DISCUSSION TITLE
DeepSpeedPlugin: non-float16 metrics

### DISCUSSION THREAD

**Discussion by leezu** [Lightning-AI/pytorch-lightning#7476] [DDP / multi-GPU / multi-node]
PL DeepSpeedPlugin passes the whole LightningModule to the DeepSpeedEngine. [DeepSpeedEngine casts the model provided to float16 during initialization](https://github.com/microsoft/DeepSpeed/blob/5b393f1555143968fbac78ecbf3c00f65baf7d78/deepspeed/runtime/engine.py#L563-L570). This includes casting the floating point states of torchmetrics associated with the module.

It would be helpful to allow using non-float16 states for metrics together with DeepSpeedPlugin.

cc @SeanNaren

**Comment by Borda** [comment]
we have introduced the float16 metrics in v0.3.0
mind check https://github.com/PyTorchLightning/metrics/releases/tag/v0.3.0

**Reply by leezu** [reply]
Thank you @Borda. I have a custom metric to represent a per-token loss where the number of tokens differs across workers. This metric goes to `inf` early in the training due to overflow if float16 states are used. It's possible to workaround with manual scaling. I'm just raising this discussion to point out that the DeepSpeedPlugin behaves differently to the other plugins

**Reply by Borda** [reply]
So you say that the inputs would be float16 but internal cumulative states float64 to overcome potential overflow, right?

**Reply by SeanNaren** [reply]
Thanks for making this issue @leezu, this is also a known issue for certain aggressive FP16 modes like APEX as well, since the entire model is kept in half precision. Might need some workaround to ensure metrics are kept in their default precision.

**Reply by leezu** [reply]
> So you say that the inputs would be float16 but internal cumulative states float64 to overcome potential overflow, right?

Yes. The inputs might even be float64 as well, given the user controls the inputs when calling the metric.

**Reply by leezu** [reply]
@Borda @SeanNaren torchmetrics 0.3 avoids the down-cast, but it's broken in version 0.4 and 0.5: https://github.com/PyTorchLightning/metrics/issues/484
